package/setup.py

```diff
@@ -130,33 +129,118 @@ def get(self, option_name, default=None):
-    """Derived class to cleanly handle setup-time (numpy) dependencies.
+class MDADistribution(Distribution, object):
+    """Derived class to cleanly handle setup-time (cython/numpy) dependencies.
     """
-    # The only setup-time numpy dependency comes when setting up its
-    #  include dir.
-    # The actual numpy import and call can be delayed until after pip
-    #  has figured it must install numpy.
-    # This is accomplished by passing the get_numpy_include function
-    #  as one of the include_dirs. This derived Extension class takes
-    #  care of calling it when needed.
-    def __init__(self, *args, **kwargs):
-        self._mda_include_dirs = []
-        super(MDAExtension, self).__init__(*args, **kwargs)
-
-    @property
-    def include_dirs(self):
-        if not self._mda_include_dirs:
-            for item in self._mda_include_dir_args:
-                try:
-                    self._mda_include_dirs.append(item()) #The numpy callable
-                except TypeError:
-                    self._mda_include_dirs.append(item)
-        return self._mda_include_dirs
-
-    @include_dirs.setter
-    def include_dirs(self, val):
-        self._mda_include_dir_args = val
+    # Setup-time cython dependency comes when cythonizing sources (cython might
+    #  not be available) or configuring numpy's include dir (numpy might
+    #  not be available either).
+    # The actual cython/numpy imports and calls can be delayed until after
+    #  pip/setuptools have figured they must install them.
+    # This is accomplished by using a setuptools.dist.Distribution subclass
+    #  that lazily fills-in cython and numpy information by hooking into
+    #  the run_commands method.
+    # Our overrides also address the setuptools behavior of installing
+    #  setup-time dependencies locally. This involves essentially a rewrite
+    #  of fetch_build_egg.
+    # Hopefully, since none of the overriden methods was pre-underscored,
+    #  we'll stay clear of problems while this setuptools API lasts.
+    def __init__(self, attrs=None):
+        # Attributes must exist if they're to be imported from the
+        #  attrs that setup feeds to __init__.
+        self.mda_use_cython = is_dev
+        self.mda_build_requires = []
+        super(MDADistribution, self).__init__(attrs)
+
+    def run_commands(self):
+        # The following might be fragile because we rely on specific
+        #  substrings of the command passed to setup.py. If pip ever starts
+        #  running a probing command other than 'egg_info' this might need to
+        #  be changed. Likewise if we ever come up with new commands ourselves
+        is_install = any(('install' in cmd for cmd in self.commands))
+        is_build = any(('build' in cmd for cmd in self.commands))
+        is_pipinstall = any(('egg_info' in cmd for cmd in self.commands))
+        if is_install or is_pipinstall:
+            # MDAnalysis can do fine with only installing dependencies at
```

Comment:
While we still need a hack to delay our own `numpy`/`cython` dependency, if we'd no longer depend on `biopython` we wouldn't need to install those dependencies ourselves at `pip` probe time.
This means that, contrary to this PR, we would not install anything on `egg_info` and, by the time `pip` runs us again with the `install` command it'll have already installed `numpy` for us. It's cleaner because then we don't need to do a mixed `setuptools`/`pip` installation as we do now.